artifact/v3/scripts/executor.cpp

```diff
@@ -177,36 +181,88 @@ void ScriptRunner::LogErrAndExecuteNext(
 		this->error_script_error_ = this->error_script_error_.FollowedBy(err);
 	}
 
-	// Schedule the next script execution
+	HandleScriptNext(current_script, end, ignore_error, handler);
+}
+
+void ScriptRunner::HandleScriptNext(
+	vector<string>::iterator current_script,
+	vector<string>::iterator end,
+	bool ignore_error,
+	HandlerFunction handler) {
+	// Stop retry timer and start the next script execution
+	if (this->retry_timeout_timer_) {
+		this->retry_timeout_timer_->Cancel();
+		this->retry_timeout_timer_.reset();
+	}
+
 	auto local_err = Execute(std::next(current_script), end, ignore_error, handler);
 	if (local_err != error::NoError) {
 		return handler(local_err);
 	}
 }
 
 void ScriptRunner::HandleScriptError(Error err, HandlerFunction handler) {
-	if (err.code
-		== processes::MakeError(processes::NonZeroExitStatusError, "").code) {
-		if (this->script_.get()->GetExitStatus() == state_script_retry_exit_code) {
-			return handler(executor::MakeError(
-				executor::RetryExitCodeError,
-				"Received exit code: " + to_string(state_script_retry_exit_code)));
-		}
+	// Stop retry timer
+	if (this->retry_timeout_timer_) {
+		this->retry_timeout_timer_->Cancel();
+		this->retry_timeout_timer_.reset();
+	}
+	if (err.code == processes::MakeError(processes::NonZeroExitStatusError, "").code) {
 		return handler(executor::MakeError(
 			executor::NonZeroExitStatusError,
 			"Received error code: " + to_string(this->script_.get()->GetExitStatus())));
 	}
 	return handler(err);
 }
 
+void ScriptRunner::HandleScriptRetry(
+	vector<string>::iterator current_script,
+	vector<string>::iterator end,
+	bool ignore_error,
+	HandlerFunction handler) {
+	log::Info(
+		"Script returned Retry Later exit code, re-retrying in "
+		+ to_string(this->retry_interval_.count() / 1000) + "s");
```

Comment:
Q: Could we not go via `chrono` somehow?

Reply:
Not really, that I could find. On the one hand the [stream output operator](https://en.cppreference.com/w/cpp/chrono/duration/operator_ltlt) is C++20 only, and on the other hand the duration conversions are only allowed from bigger to smaller unit (seconds to milliseconds, for example, but not the other way around!).

Related,

We use `milliseconds` only for the sake of tests. On production the duration unit (coming from the config file or from defaults) is seconds. Do you have a different suggestion on how could I keep using seconds in the code and only tweak it for tests?

Reply:
Alright, makes sense :)

Reply:
Could you use `chrono::duration_cast<chrono::seconds>(this->retry_interval_).count()`? It always rounds down, but I guess that isn't a big concern here.

Reply:
Oh, I missed `duration_cast`, thanks!